src/processors/reverseFull.ts

```diff
@@ -0,0 +1,223 @@
+          smartContractCommittedHeight,
+          smartContractCommittedHash,
+          mainnetBlockHash
+        );
+      }
+      return { shouldProcess: false, msg: "Smart contract committed and subnet are already in sync, nothing needs to be done, waiting for new blocks" };
+    } else {
+      // Check the committed
+      const auditedCommittedBlockInfoInMainnet =
+        await this.mainnetService.getCommittedBlockInfoByNum(
+          smartContractCommittedHeight
+        );
+      if (
+        auditedCommittedBlockInfoInMainnet.mainnetBlockHash !=
+        smartContractCommittedHash
+      ) {
+        this.logger.error(
+          "⛔️ WARNING: Forking detected when subnet is ahead of smart contract"
+        );
+        throw new ForkingError(
+          smartContractCommittedHeight,
+          smartContractCommittedHash,
+          auditedCommittedBlockInfoInMainnet.mainnetBlockHash
+        );
+      }
+      if (smartContractHash == mainnetBlockHash) {
+        // Same block height and hash
+        return { shouldProcess: false, msg: "Smart contract latest and subnet are already in sync, nothing needs to be done, waiting for new blocks" };
+      } else if (mainnetBlockNumber < smartContractHeight) {
+        const mainnetHashInSmartContract =
+          await this.mainnetService.getBlockHashByNumber(mainnetBlockNumber);
+        if ( mainnetHashInSmartContract != mainnetBlockHash) {
+          const { divergingHeight } = await this.findDivergingPoint(
+            mainnetBlockNumber
+          );
+          return {
+            shouldProcess: true,
+            from: divergingHeight,
+            msg: `Forking happened but not yet committed on mainnet, we will need to recursively submit subnet headers from diverging point of ${divergingHeight}`
+          };
+        }
+        return { 
+          shouldProcess: false,
+          msg: "Subnet is behind mainnet latest audited blocks! This usually means there is another relayer on a different node who is ahead of this relayer in terms of mining and submitting txs. OR there gonna be forking soon!"
+        };
+      }
+      // Below is the case where subnet is ahead of mainnet and we need to do some more checks before submit txs
+      const auditedBlockInfoInMainnet =
+        await this.mainnetService.getCommittedBlockInfoByNum(
+          smartContractHeight
+        );
+      if (auditedBlockInfoInMainnet.mainnetBlockHash != smartContractHash) {
+        const { divergingHeight } = await this.findDivergingPoint(
+          smartContractHeight
+        );
+        return {
+          shouldProcess: true,
+          from: divergingHeight
+        };
+      }
+      // Everything seems normal, we will just submit txs from this point onwards.
+      return {
+        shouldProcess: true,
+        from: smartContractHeight,
+      };
+    }
+  }
+  
+  // Find the point where after this "divering block", chain start to split (fork)
+  private async findDivergingPoint(
```

Comment:
quick question. is this method the same as the one we have in the subnet -> mainnet relayer job? if it is, maybe we shall consider move it into a utils file

Reply:
yes, many similarities in SubnetService and MainnetService, let's have a next PR to clean this up